### PR TITLE
[feature] Fidelity Bonds

### PR DESCRIPTION
**Pull request by rajarshimaitra**
This PR adds the Fidelity Bond API, protocol, and tests. 

The previous fidelity bond codes are mostly removed. 

The Wallet now has a new fidelity-related API.

The Makers by default attempt to create a fidelity bond and startup with default config options. If failed, then the server stops with a hard error.

Takers now validate fidelity bonds at the offer downloading time. If an invalid bond is present, that offer is rejected.

Unit and functional tests were added to test fidelity bond behavior.

Reviewer's note:
Commits are broken into modular chunks. Check out each commit message to get the change summary.

### CODE REVIEW

File: src/protocol/contract.rs [hunk 1]
```diff
@@ -10,9 +10,9 @@ use bitcoin::{
         opcodes::{self, all},
         script::{Builder, Instruction, Script},
     },
+    ecdsa::Signature,
```
Comment (Shourya742) [inline src/protocol/contract.rs:13]: `ecdsa::Signature as EcdsaSignature;
`

File: src/protocol/contract.rs [hunk 2]
```diff
@@ -10,9 +10,9 @@ use bitcoin::{
...
     hashes::Hash,
     secp256k1::{
-        ecdsa::Signature,
```
Comment (Shourya742) [inline src/protocol/contract.rs:15]: `ecdsa::Signature as SecpSignature`

File: src/protocol/contract.rs [hunk 3]
```diff
@@ -69,8 +69,8 @@ pub fn calculate_coinswap_fee(
 pub fn apply_two_signatures_to_2of2_multisig_spend(
     key1: &PublicKey,
     key2: &PublicKey,
-    sig1: &Signature,
-    sig2: &Signature,
+    sig1: &bitcoin::secp256k1::ecdsa::Signature,
```
Comment (Shourya742) [inline src/protocol/contract.rs:72]: `sig1: &SecpSignature`

File: src/protocol/contract.rs [hunk 4]
```diff
@@ -69,8 +69,8 @@ pub fn calculate_coinswap_fee(
...
+    sig2: &bitcoin::secp256k1::ecdsa::Signature,
```
Comment (Shourya742) [inline src/protocol/contract.rs:73]: `sig2: &SecpSignature`

File: src/protocol/contract.rs [hunk 5]
```diff
@@ -489,7 +493,7 @@ pub fn verify_contract_tx_sig(
     multisig_redeemscript: &Script,
     funding_amount: u64,
     pubkey: &PublicKey,
-    sig: &Signature,
+    sig: &bitcoin::secp256k1::ecdsa::Signature,
```
Comment (Shourya742) [inline src/protocol/contract.rs:496]: `sig: &SecpSignature`

**Comment by rajarshimaitra** [comment]
@Shourya742 removed secp signatures entirely.. No reason we needed both..

File: src/wallet/fidelity.rs [hunk 6]
```diff
@@ -15,456 +38,582 @@ const BOND_VALUE_EXPONENT: f64 = 1.3;
...
+        let (index, fidelity_addr, fidelity_pubkey) = self.get_next_fidelity_address(locktime)?;
+
+        // Fetch utxos, filter out existing fidelity coins
+        let mut unspents = self
+            .list_unspent_from_wallet(false, false)?
+            .into_iter()
+            .filter(|(_, spend_info)| !matches!(spend_info, UTXOSpendInfo::FidelityBondCoin { .. }))
+            .collect::<Vec<_>>();
+
+        unspents.sort_by(|a, b| b.0.amount.cmp(&a.0.amount));
```
Comment (Shourya742) [inline src/wallet/fidelity.rs:319]: Any specific reason to sort in desc,? As both sorted and non-sorted unspents will result in sub optimal solutions. Do we want to have selected utxo size to be as low as possible?

File: src/wallet/fidelity.rs [hunk 7]
```diff
@@ -15,456 +38,582 @@ const BOND_VALUE_EXPONENT: f64 = 1.3;
...
+
+        // the simplest largest first coinselection.
+        for unspent in unspents {
+            if remaining.checked_sub(unspent.0.amount).is_none() {
+                selected_utxo.push(unspent);
+                break;
+            } else {
+                remaining -= unspent.0.amount;
+                selected_utxo.push(unspent);
+                continue;
```
Comment (Shourya742) [inline src/wallet/fidelity.rs:332]: continue can be removed here:
``` 
else {
                remaining -= unspent.0.amount;
                selected_utxo.push(unspent);
     }
```

File: src/wallet/fidelity.rs (hunk 6 above)
Comment (rajarshimaitra) [reply src/wallet/fidelity.rs:319]: For the dumb coinselection I did, it's easier to write it if it's sorted.. Its suboptimal anyway.. We won't be targeting wallet optimization here anyway.. 

And yes the sort allow us to have lowest selection size. Not that its super important.. 

corollary idea: having an independent rust coinselection library would be super awesome..

File: src/wallet/fidelity.rs (hunk 7 above)
Comment (rajarshimaitra) [reply src/wallet/fidelity.rs:332]: Makes sense..

**Review by wthrajat — APPROVED** [review]

File: tests/malice2.rs [hunk 8]
```diff
@@ -132,7 +145,7 @@ async fn malice2_maker_broadcast_contract_prematurely() {
         .balance(false, false)
         .unwrap();
 
-    assert_eq!(maker_balances.first().unwrap(), &Amount::from_sat(14995773));
+    assert_eq!(maker_balances.first().unwrap(), &Amount::from_sat(14994773));
 
     // Everybody looses 4227 sats for contract transactions.
```
[the remark this answers is not in the document]
Comment (rajarshimaitra) [reply tests/malice2.rs:150]: Not really. The `org_maker_balances` is also reduced by 1000 sats. Thats the fees paid for fidelity tx. so the diff remains the same.

File: src/maker/config.rs [hunk 9]
```diff
@@ -48,6 +52,8 @@ impl Default for MakerConfig {
             required_confirms: 1,
             min_contract_reaction_time: 48,
             min_size: 10_000,
+            fidelity_value: 5_000_000, // 50 million  sats
```
[the remark this answers is not in the document]
Comment (rajarshimaitra) [reply src/maker/config.rs:55]: good catch..

**Pull request merged**